ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll

```diff
@@ -0,0 +1,423 @@
+ * GraphQL `field`.
+ *
+ * For example:
+ *
+ * ```rb
+ * class Types::User < GraphQL::Schema::Object
+ *   field :email, String
+ *   field :friends, [Types::User] do
+ *     argument :starts_with, String, "Show only friends matching the given prefix"
+ *   end
+ *
+ *   # GraphqlFieldResolutionMethod
+ *   def friends(starts_with:)
+ *     object.friends.where("name like #{starts_with}")
+ *   end
+ * end
+ * ```
+ *
+ * or:
+ *
+ * ```rb
+ * class Types::User < GraphQL::Schema::Object
+ *   field :email, String
+ *   field :friends, [Types::User] do
+ *     argument :starts_with, String, "Show only friends matching the given prefix",
+ *       resolver_method: :my_custom_method, extras: [:graphql_name]
+ *   end
+ *
+ *   # GraphqlFieldResolutionMethod
+ *   def my_custom_method(**args)
+ *     puts args[:graphql_name] # for debugging
+ *     object.friends.where("name like #{args[:starts_with]}")
+ *   end
+ * end
+ * ```
+ */
+class GraphqlFieldResolutionMethod extends Method, HTTP::Server::RequestHandler::Range {
+  private GraphqlSchemaObjectClass schemaObjectClass;
+
+  GraphqlFieldResolutionMethod() {
+    this.getEnclosingModule() = schemaObjectClass and
+    exists(GraphqlFieldDefinitionMethodCall defn |
+      // field :foo, resolver_method: :custom_method
+      // def custom_method(...)
+      defn.getKeywordArgument("resolver_method").(SymbolLiteral).getValueText() = this.getName()
+      or
+      // field :foo
+      // def foo(...)
+      not exists(defn.getKeywordArgument("resolver_method").(SymbolLiteral)) and
+      defn.getFieldName() = this.getName()
+    )
+  }
+
+  /** Gets the method call which is the definition of the field corresponding to this resolver method. */
+  GraphqlFieldDefinitionMethodCall getDefinition() {
+    result.getKeywordArgument("resolver_method").(SymbolLiteral).getValueText() = this.getName()
+    or
+    not exists(result.getKeywordArgument("resolver_method").(SymbolLiteral)) and
+    result.getFieldName() = this.getName()
+  }
+
+  // check for a named argument the same name as a defined argument for this field
+  override Parameter getARoutedParameter() {
+    result = this.getAParameter() and
+    exists(GraphqlFieldArgumentDefinitionMethodCall argDefn |
+      argDefn.getEnclosingCallable() = this.getDefinition().getBlock() and
+      (
+        result.(NamedParameter).hasName(argDefn.getArgumentName())
+        or
+        not result.(NamedParameter).hasName(argDefn.getArgumentName()) and
+        // TODO this will cause false positives because now *anything* in the **args
+        // param will be flagged as as RoutedParameter/RemoteFlowSource, but really
+        // only the hash keys corresponding to the defined arguments are user input
+        // others could be things defined in the `:extras` keyword argument to the `argument`
+        exists(result.(HashSplatParameter)) // often you see `def field(**args)`
```

Comment:
This is probably _slightly_ more idiomatic as `result instanceof HashSplatParameter`.